core/base/segmented_array.cpp

```diff
@@ -0,0 +1,89 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <ginkgo/core/base/segmented_array.hpp>
+
+
+#include "core/base/array_access.hpp"
+
+
+namespace gko {
+
+
+template <typename T>
+segmented_array<T>::segmented_array(std::shared_ptr<const Executor> exec)
+    : buffer_(exec), offsets_(exec)
+{}
+
+
+template <typename T>
+segmented_array<T>::segmented_array(std::shared_ptr<const Executor> exec,
+                                    const std::vector<int64>& sizes)
+    : segmented_array(
+          gko::array<T>(exec,
+                        std::accumulate(sizes.begin(), sizes.end(), int64{})),
+          sizes)
+{}
+
+
+template <typename T>
+segmented_array<T>::segmented_array(array<T> buffer,
+                                    const std::vector<int64>& sizes)
+    : buffer_(std::move(buffer))
+{
+    auto exec = this->get_executor();
+    offsets_ = gko::array<int64>(exec->get_master(), sizes.size() + 1);
+    offsets_.fill(0);
+    std::partial_sum(sizes.begin(), sizes.end(), offsets_.get_data() + 1);
+    GKO_ASSERT_EQ(buffer_.get_size(), offsets_.get_data()[sizes.size()]);
+    offsets_.set_executor(exec);
+}
+
+
+template <typename T>
+segmented_array<T>::segmented_array(array<T> buffer, gko::array<int64> offsets)
+    : buffer_(std::move(buffer)), offsets_(std::move(offsets))
+{
+    offsets_.set_executor(this->get_executor());
+    GKO_ASSERT_EQ(buffer_.get_size(),
+                  offsets_.get_size()
```

Comment:
offsets_.size() should never be 0

Reply:
I was also thinking that. I decided against that, because then I would need to allocate a new array during a move. Do you think that would be acceptable?

Reply:
Our moves are mostly not `noexcept` anyways, because we can have cross-executor copies when move-assigning. Performance-wise, allocating <= 8 bytes shouldn't really matter, unless we do really weird things.